core/src/main/java/org/ldaptive/pool/AbstractConnectionPool.java

```diff
@@ -1061,56 +1064,55 @@ public void prune()
     poolLock.lock();
     try {
       if (!available.isEmpty()) {
-        final int currentPoolSize = active.size() + available.size();
-        if (currentPoolSize > minPoolSize) {
-          logger.debug("Pruning available pool of size {} for {}", available.size(), this);
-
-          final int numConnAboveMin = currentPoolSize - minPoolSize;
-          final int numConnToPrune = available.size() < numConnAboveMin ? available.size() : numConnAboveMin;
-          final List<Callable<PooledConnectionProxy>> callables = new ArrayList<>(numConnToPrune);
-          for (PooledConnectionProxy pc : available) {
-            logger.trace("pruning {} for {}", pc, this);
-            callables.add(() -> {
-              if (pruneStrategy.apply(pc)) {
-                logger.trace("prune approved on {} with {} for {}", pc, pruneStrategy, AbstractConnectionPool.this);
-                return pc;
+        final CallableWorker<Function<PrunePoolState, PooledConnectionProxy>> callableWorker =
+          new CallableWorker<>(name + "-prune");
+        final AtomicInteger numConnPruned = new AtomicInteger();
+        try {
+          final int size = pruneStrategy.getPredicateSize();
+          logger.trace("prune strategy has {} predicates", size);
+          // prune all available connections for each predicate supplied by the prune strategy
+          IntStream.range(0, size)
+            .forEach(i -> {
+              final List<Callable<Function<PrunePoolState, PooledConnectionProxy>>> callables =
+                new ArrayList<>();
+              for (PooledConnectionProxy pc : available) {
+                callables.add(() -> poolState -> {
+                  if (pruneStrategy.apply(pc).get(i).test(poolState))
```

Comment:
The reason for the list of predicates is so each predicate can operate over the _entire pool_ before the next predicate executes. So `isIdle(a, b)` cannot be evaluated until `super.apply(a, b)` has been applied to every connection in the pool. (The list semantics implies priority) But I'm certainly open to alternatives.

I would add that the complexity in this API is a direct result of the decision to couple prune configuration properties to the strategy. (prunePeriod, statisticsSize, etc). Due to that decision the pool must have a single `PruneStrategy`. Otherwise the solution is simply to set a list of prune strategies on the pool and I think the effects are easy to reason about. That said, I wouldn't change the API to support that behavior. Configuring a single prune strategy has other benefits.